### PR TITLE
Js task 3

### PR DESCRIPTION
**Pull request by dudzon**
https://dudzon.github.io/ui-internship-lwo2/css_tasks/task10/

### CODE REVIEW

File: css_tasks/task10/index.html [hunk 1]
```diff
@@ -261,25 +261,35 @@ <h2 class="header__title header__title--center header__title--form">
...
-          name="email"
-        />
-        <input
-          class="inner-container__element inner-container__element--form"
-          type="password"
-          placeholder="password"
-          name="password"
-        />
+      <form
+        action=""
```
Comment (tkachukivan) [inline css_tasks/task10/index.html]: at least post to not see all typed stuff at url, all some good ux with cleaning form fields and showing some message to user that form was sended

File: css_tasks/task10/js/app.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+// EVENTS
+signupForm.addEventListener("submit", submitForm);
+inputs.forEach(item =>
+  item.addEventListener("focus", () =>
+    removeDisabledState(formButton, formText)
+  )
+);
+signupForm.addEventListener("submit", () => addDisabledState(formButton));
+formButton.addEventListener("click", () => hideFormText(formText));
```
Comment (tkachukivan) [inline css_tasks/task10/js/app.js]: here you create 3 listeners in place where it possible to create 1

what if user press enter to submit form and not click on button?

File: css_tasks/task10/js/constants.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable max-len */
+export const constants = {
+  nameText: 'Only latin letters',
+  namePattern: /^[a-zA-Z ]+$/,
+  emailText: 'Provide valid email address',
+  emailPattern: /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
+  passwordText:
+    'Password should have at least  6 characters including:capital letter, lowercased letter, digit and special character',
+  passwordPattern: /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{6,}$/,
+  formButton: document.querySelector('.form__button'),
```
Comment (tkachukivan) [inline css_tasks/task10/js/constants.js]: usually in constant some values(e.g primitives ) which not changed but not dom elements

File: css_tasks/task10/js/form.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+// Add disabled class to the button
+
+const addDisabledState = (elem) => {
+  const inputContainers = [...formInputContainer];
+  inputContainers.forEach((item) => {
+    const lastChild = item.lastElementChild.classList.contains('form-error');
+    if (lastChild) {
+      elem.classList.add('disabled');
```
Comment (tkachukivan) [inline css_tasks/task10/js/form.js]: hm, buttons have attribute disabled, no reason for separate class here

**Review by tkachukivan — CHANGES_REQUESTED** [review]
disabled button almost not visible, looks like it disappear

File: css_tasks/task10/js/constants.js (hunk 3 above)
Comment (dudzon) [reply css_tasks/task10/js/constants.js]: Changed

File: css_tasks/task10/js/form.js (hunk 4 above)
Comment (dudzon) [reply css_tasks/task10/js/form.js]: Changed

File: css_tasks/task10/index.html (hunk 1 above)
Comment (dudzon) [reply css_tasks/task10/index.html]: Changed

File: css_tasks/task10/js/app.js (hunk 2 above)
Comment (dudzon) [reply css_tasks/task10/js/app.js]: Actually it worked the same, but I get your point

Comment (dudzon) [reply css_tasks/task10/js/app.js]: Changed

File: css_tasks/task10/js/form.js [hunk 5]
```diff
@@ -70,26 +78,80 @@ const hideWarning = (elem) => {
...
 const hideFormText = (elem) => {
   elem.style.display = 'none';
 };
-export {submitForm, addDisabledState, removeDisabledState, hideFormText};
+
+//  Create success message
+const createSuccessMessage = () => {
+  const successMessage = document
+      .createRange()
+      .createContextualFragment(
```
Comment (tkachukivan) [inline css_tasks/task10/js/form.js]: it is possible to do simpler with older methods

**Review by tkachukivan — APPROVED** [review]

File: css_tasks/task10/js/app.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import {submitForm} from './form.js';
+import {DOMElements} from './elements.js';
+const {signupForm} = DOMElements;
+
+// EVENTS
+signupForm.addEventListener('submit', submitForm);
```
Comment (mikhail-ev) [inline css_tasks/task10/js/app.js]: any access to your DOM should be wrapped to DOMContentLoaded callback.
Not each one, but the whole script that's going to access DOM

File: css_tasks/task10/js/form.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  hideWarning(email);
+  email.addEventListener('focus', () => removeDisabledState(formButton));
+  if (!emailPattern.test(emailValue)) {
+    handleError(email, emailText, formButton);
+  }
+};
+
+// Validate Password Input
+
+const validatePassword = (formName) => {
```
Comment (mikhail-ev) [inline css_tasks/task10/js/form.js]: all 3 validation functions can be turned into 1 that takes a validation callback of any form of delegation.

File: css_tasks/task10/js/form.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    handleError(password, passwordText, formButton);
+  }
+};
+
+// Show warning
+
+const showWarning = (formName, text) => {
+  const warningMessage = document
+      .createRange()
+      .createContextualFragment(`<p class="form-error">${text}</p>`);
```
Comment (mikhail-ev) [inline css_tasks/task10/js/form.js]: why are you using range instead of createElement?

File: css_tasks/task10/js/form.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+const hideWarning = (elem) => {
+  const warning = elem.nextElementSibling;
+  if (warning && warning.classList.contains('form-error')) {
+    warning.remove();
+  }
+};
+
+// Add disabled attribute to the button
+
+const addDisabledState = (elem) => {
```
Comment (mikhail-ev) [inline css_tasks/task10/js/form.js]: probably you shouldn't rely on class on *another* element, it's unintuitive and required traversing each time. You can store some info about element using data attributes

File: css_tasks/task10/js/form.js [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+const handleError = (input, inputText, handler) => {
+  showWarning(input, inputText);
+  addDisabledState(handler);
+  hideFormText(formText);
+};
+
+//  Check if there are any errors displayed
+
+const checkErrors = () => {
```
Comment (mikhail-ev) [inline css_tasks/task10/js/form.js]: same as above, don't rely on other elements, store data on data attributes or any other way but on target element itself

**Review by mikhail-ev — CHANGES_REQUESTED** [review]

**Pull request merged**